platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetScoreCalculator.java

```diff
@@ -0,0 +1,281 @@
+     *
+     * @return the current tipset snapshot
+     */
+    public @NonNull Tipset getSnapshot() {
+        return snapshot;
+    }
+
+    /**
+     * Add an event created by this node and compute the increase in tipset score. Higher score changes mean that this
+     * event caused consensus to advance more. A score change of 0 means that this event did not advance consensus. A
+     * score change close to the total weight means that this event did a very good job at advancing consensus. It's
+     * impossible to get a perfect score, since the weight of advancing self events is not included. The maximum score
+     * an event can achieve is equal to the sum of all weights minus this node's weight.
+     * <p>
+     * Whenever the total advancement score of a new event exceeds the threshold (2/3 minus self weight), the snapshot
+     * is set to be equal to this event's tipset.
+     *
+     * @param event the event that is being added
+     * @return the change in this event's tipset score compared to the tipset score of the previous event passed to this
+     * method
+     */
+    public TipsetAdvancementWeight addEventAndGetAdvancementScore(@NonNull final EventDescriptor event) {
+        Objects.requireNonNull(event);
+        if (!event.getCreator().equals(selfId)) {
+            throw new IllegalArgumentException("event creator must be the same as the window ID");
+        }
+
+        final Tipset eventTipset = tipsetTracker.getTipset(event);
+        if (eventTipset == null) {
+            throw new IllegalArgumentException("event " + event + " is not in the tipset tracker");
+        }
+
+        final TipsetAdvancementWeight score = snapshot.getTipAdvancementWeight(selfId, eventTipset);
+        if (score.advancementWeight() > maximumPossibleAdvancementWeight) {
+            throw new IllegalStateException("score " + score + " is greater than the maximum possible score "
+                    + maximumPossibleAdvancementWeight);
+        }
+
+        final TipsetAdvancementWeight scoreImprovement = score.minus(previousScore);
+
+        if (isSuperMajority(score.advancementWeight() + selfWeight, totalWeight)) {
+            snapshot = eventTipset;
+            snapshotHistory.add(snapshot);
+            if (snapshotHistory.size() > maxSnapshotHistorySize) {
+                snapshotHistory.remove();
+            }
+            previousScore = ZERO_ADVANCEMENT_WEIGHT;
+        } else {
+            previousScore = score;
+        }
+
+        latestSelfEventTipset = eventTipset;
+
+        return scoreImprovement;
+    }
+
+    /**
+     * Figure out what advancement score we would get if we created an event with a given list of parents.
+     *
+     * @param parents the proposed parents of an event
+     * @return the advancement score we would get by creating an event with the given parents
+     */
+    public TipsetAdvancementWeight getTheoreticalAdvancementScore(@NonNull final List<EventDescriptor> parents) {
```

Comment:
Is the parameter a list in anticipation of future features?

Reply:
Yes, it's in anticipation of supporting multiple other parents. Once the hashgraph is capable of supporting multiple other parents, it will be a very minor change to this code to start generating events with many parents.